AssemblyScript ports of ecosystem contracts

mpapierski opened the pull request:
Closes product#206

Port of hello world https://github.com/casper-ecosystem/hello-world, and counter https://github.com/casper-ecosystem/counter to AssemblyScript.

This PR also brings up the AssemblyScript compiler to 0.17 (currently the newest is 0.20.x). Interestingly `asc` 0.18+ always emits a start section which we now don't support  (#2485), and I haven't found a straightforward way to disable it—still, a pretty good improvement over the previously supported 0.10.

Commit:
Bump assemblyscript compiler to 0.17

Commit:
Add assemblyscript ports of contracts.

MaxGraey commented:
AssemblyScript 0.20.x has an option to redefine `start` and export custom function instead. See `--exportStart` option in docs: https://www.assemblyscript.org/compiler.html#compiler-options

Usage: `asc ... --exportStart customStart` will generate:
```wat
(module
 (type $none_=>_none (func))
 (memory $0 0)
 (export "memory" (memory $0))
 (export "customStart" (func $~customStart))
 (func $~customStart
  ...
 )
)
```

mpapierski commented:
> AssemblyScript 0.20.x has an option to redefine `start` and export custom function instead. See `--exportStart` option in docs: https://www.assemblyscript.org/compiler.html#compiler-options
> 
> Usage: `asc ... --exportStart customStart` will generate:
> 
> ```webassembly
> (module
>  (type $none_=>_none (func))
>  (memory $0 0)
>  (export "memory" (memory $0))
>  (export "customStart" (func $~customStart))
>  (func $~customStart
>   ...
>  )
> )
> ```

Thanks for reaching out, @MaxGraey, unfortunately `--exportStart` produces a strange `AS110` error at compilation time:

```sh
$ asc --version
Version 0.20.12
$ asc --lib ../../.. assembly/index.ts -o ../../../../target_as/counter_define.wasm --exportStart --disable bulk-memory --optimize --use abort=

ERROR AS110: Start function name '' is invalid or conflicts with another export.

 [native code]
 ~~~~~~~~~~~~~
 in ~lib/native.ts(1,1)

FAILURE 1 compile error(s)
```

Not having a start section is an improvement, although if I understand correctly, our host interface still would have to know about exported start function and call it externally. I think we're better off supporting the start section, but it would still be interesting to know what changed in 0.18 that our code suddenly generates a start section, but in 0.17 doesn't? Is it the global variables that are causing it?

On a side note assuming `--exportStart` would work, would it be possible to use the `--transform` feature to put a `(call $~customStart)` at the beginning of an exported `call` function?

MaxGraey commented:
You should use it as `--exportStart <your-some-custom-start-name>` (for example `--exportStart  _start`) instead just single `--exportStart`. It's option with parameter

Commit:
Fix flags

mpapierski commented:
> You should use it as `--exportStart <your-some-custom-start-name>` (for example `--exportStart _start`) instead just single `--exportStart`. It's option with parameter

Right, now this flag is working as intended. Some kind of readable error message would be helpful if an argument is missed though. Anyway thanks @MaxGraey for help!

Commit:
Update to the structure

- Create counter-installer for rust.
- Create increment-counter session for rust.
- Rename assemblyscript ports accordingly.
- Create hello-world equivalent for rust.
- Simplify testing of tutorial contracts.

darthsiroftardis commented on smart_contracts/contracts/tutorial/counter-installer/src/main.rs:
```diff
@@ -0,0 +1,97 @@
+    let version_uref = storage::new_uref(contract_version);
+    runtime::put_key(CONTRACT_VERSION_KEY, version_uref.into());
+
+    // To create a locked contract instead, use new_locked_contract and throw away the contract
```
Can we move these lines to above the `put_key` please?

mpapierski replied:
Done

darthsiroftardis commented on smart_contracts/contracts_as/client/withdraw-bid/assembly/index.ts:
```diff
@@ -1,17 +1,18 @@
+
 
 export function call(): void {
+    const ARG_AMOUNT = "amount";
```
This seems to be the only place constants are being defined within the function body. Is there a particular reason for it?

mpapierski replied:
I experimented with the newest AssemblyScript version, and I thought global constants like this suddenly started to generate a wasm `start` section, which we don't currently support, but it's unrelated to the issue. I moved it above as it should be.

darthsiroftardis commented on smart_contracts/contracts_as/test/do-nothing-stored/assembly/index.ts:
```diff
@@ -32,9 +32,9 @@ export function call(): void {
     ACCESS_KEY_NAME,
   );
   const key = Key.create(CLValue.fromI32(result.contractVersion));
```
Can we rename this variable, @Fraser999 has convinced me its best to avoid having the variable name be the same as the type itself

darthsiroftardis commented on smart_contracts/contracts_as/test/groups/assembly/index.ts:
```diff
@@ -113,9 +113,9 @@ export function call_restricted_entry_points(): void {
 
 function createGroup(packageHash: Uint8Array): URef {
   let key = Key.create(CLValue.fromU64(0));
```
Same here about the variable rename

mpapierski replied:
Good point - I did a full pass on other contracts as well in c6276b99e5574cb0ec38075f99b9472707b84877

darthsiroftardis reviewed (approved):
Some overall nits, but looks good

Commit:
Apply @darthsiroftardis comments

mpapierski commented:
bors r+
